accounts-on-demand/src/accounts_on_demand.rs

```diff
@@ -0,0 +1,228 @@
+        account_filter::{AccountFilter, AccountFilterType, AccountFilters},
+    },
+};
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use tokio::sync::{broadcast::Sender, RwLock};
+
+use crate::subscription_manager::SubscriptionManger;
+
+pub struct AccountsOnDemand {
+    rpc_client: Arc<RpcClient>,
+    accounts_storage: Arc<dyn AccountStorageInterface>,
+    accounts_subscribed: Arc<DashSet<Pubkey>>,
+    program_filters: Arc<RwLock<AccountFilters>>,
+    subscription_manager: SubscriptionManger,
+}
+
+impl AccountsOnDemand {
+    pub fn new(
+        rpc_client: Arc<RpcClient>,
+        grpc_sources: Vec<GrpcSourceConfig>,
+        accounts_storage: Arc<dyn AccountStorageInterface>,
+        account_notification_sender: Sender<AccountNotificationMessage>,
+    ) -> Self {
+        Self {
+            rpc_client,
+            accounts_storage: accounts_storage.clone(),
+            accounts_subscribed: Arc::new(DashSet::new()),
+            program_filters: Arc::new(RwLock::new(vec![])),
+            subscription_manager: SubscriptionManger::new(
+                grpc_sources,
+                accounts_storage,
+                account_notification_sender,
+            ),
+        }
+    }
+
+    pub async fn reset_subscription(&self) {
+        let mut filters = self.get_filters().await;
+
+        // add additional filters related to accounts
+        for accounts in &self
+            .accounts_subscribed
+            .iter()
+            .map(|x| x.to_string())
+            .chunks(100)
+        {
+            let account_filter = AccountFilter {
+                accounts: accounts.collect(),
+                program_id: None,
+                filters: None,
+            };
+            filters.push(account_filter);
+        }
+
+        self.subscription_manager
+            .update_subscriptions(filters)
+            .await;
+    }
+
+    pub async fn get_filters(&self) -> AccountFilters {
+        self.program_filters.read().await.clone()
+    }
+}
+
+#[async_trait]
+impl AccountStorageInterface for AccountsOnDemand {
+    async fn update_account(&self, account_data: AccountData, commitment: Commitment) -> bool {
+        self.accounts_storage
+            .update_account(account_data, commitment)
+            .await
+    }
+
+    async fn initilize_account(&self, account_data: AccountData) {
+        self.accounts_storage.initilize_account(account_data).await
+    }
+
+    async fn get_account(&self, account_pk: Pubkey, commitment: Commitment) -> Option<AccountData> {
+        match self
+            .accounts_storage
+            .get_account(account_pk, commitment)
+            .await
+        {
+            Some(account_data) => Some(account_data),
+            None => {
+                // account does not exist in account store
+                // first check if we have already subscribed to the required account
+                if !self.accounts_subscribed.contains(&account_pk) {
```

Comment:
IMO this is a race condition.
the resubscription happens asynchronously - i.e. the values in `accounts_subscribed` are "subject to subscription" but might not yet arrived in the store. as you skip the RPC lookup in this case the account appears as if it does not exist

Reply:
account_subscribed does not contain the value we add it and reset subscription. This is done because there might be accounts that do not exists so we want to check if these accounts were already subscribed if not then we start get it from RPC and restart the subscription.